### PR TITLE
Set priority for eth transactions

### PR DESCRIPTION
**Pull request by yihuang**
## Description

Set the tx priority to the lowest priority in the messages.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -52,6 +52,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (feemarket) [\#1165](https://github.com/evmos/ethermint/pull/1165) Add hint in specs about different gas terminology for gas in Cosmos and Ethereum.
 * (rpc) [\#1169](https://github.com/evmos/ethermint/pull/1169) Remove unnecessary queries from `getBlockNumber` function
 * (ante) [1208](https://github.com/evmos/ethermint/pull/1208) Change default `maxGasWanted` value 
+* (ante) []() Set mempool priority to evm transactions.
```
Comment (yihuang) [inline CHANGELOG.md:55]:
```suggestion
* (ante) [#1214](https://github.com/evmos/ethermint/pull/1214) Set mempool priority to evm transactions.
```

**Comment by facs95** [comment]
Unit tests are failing @yihuang 🙏

File: x/evm/keeper/utils.go [hunk 2]
```diff
@@ -48,41 +55,53 @@ func (k Keeper) DeductTxCostsFromUserBalance(
...
 	}
-	return fees, nil
+
+	// calculate priority based on effective gas price
+	tipPrice := txData.EffectiveGasPrice(baseFee)
+	// if london hardfork is not enabled, tipPrice is the gasPrice
+	if baseFee != nil {
+		tipPrice = new(big.Int).Sub(tipPrice, baseFee)
+	}
+	priorityBig := new(big.Int).Quo(tipPrice, DefaultPriorityReduction.BigInt())
```
Comment (fedekunze) [inline x/evm/keeper/utils.go:97]: would be good to add tests with different priorities

**Review by fedekunze — COMMENTED** [review]
LGTM, would be great to add more integration tests and update the specification. Some remarks

1. I see the case for some applications that might not want to enable the prioritized mempool so in that scenario, we should add a parameter to the EVM to disable tx prioritization
2. Do we need to update the Cosmos ante handler to enable prioritization?

**Comment by yihuang** [comment]
> * I see the case for some applications that might not want to enable the prioritized mempool so in that scenario, we should add a parameter to the EVM to disable tx prioritization

The validator is still free to choose the `v0` mempool which will ignore the priority field.

> * Do we need to update the Cosmos ante handler to enable prioritization?

The cosmos ante handler has a default mechanism to extract prioritization from fees directly, in another PR we'll customize that one with feemarket.

File: x/evm/keeper/utils.go [hunk 3]
```diff
@@ -14,20 +15,26 @@ import (
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
 )
 
+// DefaultPriorityReduction is the default amount of price values required for 1 unit of priority.
+// Because priority is `int64` while price is `big.Int`, it's necessary to scale down the range to keep it more pratical.
+// The default value is the same as the `sdk.DefaultPowerReduction`.
+var DefaultPriorityReduction = sdk.DefaultPowerReduction
```
Comment (fedekunze) [inline x/evm/keeper/utils.go:21]: still not sure why this needs to be a separate variable

**Review by fedekunze — CHANGES_REQUESTED** [review]
We should add specific integration tests for tx prioritization before we can merge this:

- legacy and access list txs
- dynamic fee tx with 0 and  > 0 priority tip

We also need to update the specification for the EVM with more details about tx prioritization for Ethereum msgs

File: x/evm/keeper/utils.go (hunk 3 above)
Comment (yihuang) [reply x/evm/keeper/utils.go:21]: Do you mean to use `sdk.DefaultPowerReduction` directly? I think they serve a different purpose, the user may want to scale to different degree?

**Comment by yihuang** [comment]
> We should add specific integration tests for tx prioritization before we can merge this:
> 
> * legacy and access list txs
> * dynamic fee tx with 0 and  > 0 priority tip

added a priority integration test, which checked that a higher priority tx gets included first even if sent later.

> 
> We also need to update the specification for the EVM with more details about tx prioritization for Ethereum msgs

added sth in feemarket's spec in the ante handlers section.

File: x/evm/keeper/utils.go [hunk 4]
```diff
@@ -14,20 +15,26 @@ import (
...
 // DeductTxCostsFromUserBalance it calculates the tx costs and deducts the fees
+// returns (effectiveFee, priority, error)
 func (k Keeper) DeductTxCostsFromUserBalance(
 	ctx sdk.Context,
 	msgEthTx evmtypes.MsgEthereumTx,
 	txData evmtypes.TxData,
 	denom string,
 	homestead, istanbul, london bool,
-) (sdk.Coins, error) {
+) (sdk.Coins, int64, error) {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/utils.go]: done

File: go.mod [hunk 5]
```diff
@@ -191,4 +191,7 @@ replace (
 	// TODO Remove it: https://github.com/cosmos/cosmos-sdk/issues/10409
 	github.com/gin-gonic/gin => github.com/gin-gonic/gin v1.7.0
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+
+	// TODO: remove after fixed https://github.com/cosmos/cosmos-sdk/issues/11364
+	github.com/zondax/hid => github.com/zondax/hid v0.9.0
```
Comment (fedekunze) [inline go.mod:196]: why do we need this here?

File: tests/integration_tests/test_priority.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    # send the txs from low priority to high,
+    # but the later sent txs should be included earlier.
+    txhashes = [w3.eth.send_raw_transaction(tx.rawTransaction) for tx in signed]
+
+    receipts = [w3.eth.wait_for_transaction_receipt(txhash) for txhash in txhashes]
+    print(receipts)
+    assert all(receipt.status == 1 for receipt in receipts), "expect all txs success"
+
+    # the later txs should be included earlier because of higher priority
+    # FIXME there's some non-deterministics due to mempool logic
```
Comment (fedekunze) [inline tests/integration_tests/test_priority.py:103]: this should be addressed before merging this

File: tests/integration_tests/test_priority.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    priorities = [tx_priority(tx, base_fee) for _, tx in test_cases]
+    assert all(a < b for a, b in zip(priorities, priorities[1:]))
+
+    signed = [sign_transaction(w3, tx, key=KEYS[sender]) for sender, tx in test_cases]
+    # send the txs from low priority to high,
+    # but the later sent txs should be included earlier.
+    txhashes = [w3.eth.send_raw_transaction(tx.rawTransaction) for tx in signed]
+
+    receipts = [w3.eth.wait_for_transaction_receipt(txhash) for txhash in txhashes]
+    print(receipts)
```
Comment (yihuang) [inline tests/integration_tests/test_priority.py:99]:
```suggestion
```

**Review by fedekunze — APPROVED** [review]
ACK. @facs95 @alexanderbez can you review?

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**